### ISSUE TITLE
Denizen Language Basics

### ISSUE THREAD

**Issue by Morphan1** [MusicScore/Denizen-Text-Guides#1]
## Introduction
As Denizen has grown it has become more complex, and usually newer features are decently documented in respect to users that are already knowledgeable about the very basics of the language. However, to my knowledge there is almost no documentation explaining the basics in any amount of detail.

## Existing "Basics" Documentation
This is sadly the best written explanation available for what a script is:
https://one.denizenscript.com/denizen/lngs/script
This is probably the most complicated concept to explain, which is likely why existing documentation skips over it entirely, instead opting to teach by example. If an explanation could be crafted, that would be fantastic.

Here is an explanation describing command syntax as provided by meta documentation:
https://one.denizenscript.com/denizen/lngs/command%20syntax
However, I don't believe there is anything explaining what a command *is*.

Here is an attempt at explaining what `dObjects` (now `ObjectTags`) are, and detailing the specifics of each core type:
https://one.denizenscript.com/denizen/lngs/objecttags
Note that this explanation is a bit unclear, and since the replacement of `p@player` to `PlayerTag` it has become even more difficult due to nonsensical and unexplained snippets such as:
```
Why <player.name> and not <PlayerTag.name>?
```
## Suggestions
Perhaps it isn't the best, but I have a small suggestion on how commands could be explained. There is a high chance that using the word "command" could confuse users who come from a background of playing on or hosting Minecraft servers. Ideally, there should be an explanation of what a Denizen command is and what a Denizen command is not:

### What a command *is not*
Denizen commands are not Minecraft or Bukkit/Spigot commands. They have similarities, but Minecraft commands often perform a single action and are limited to one-time runs.

### What a command *is*
Denizen commands, in contrast, tend to act more like *individual instructions that come together* to create scripts. They range anywhere from sending text to a player (`narrate`) to delaying following commands (`wait`) to running other commands multiple times, possibly forever (`while`). 

Denizen commands are puzzle pieces that can be put together to change anything and everything about how a server works. Minecraft commands are best used for one-time small changes to a server.

## In conclusion
These are arguably some of the most important concepts in Denizen. We should be able to point out specific details about them to new users. Otherwise, we're essentially demanding all users to absorb the information purely through osmosis.

Aside from your own mind, I'd recommend talking to some of the other experienced users for input on these topics and how to possibly best explain them.

**Comment by mcmonkey4eva** [comment]
`Note that this explanation is a bit unclear, and since the replacement of p@player to PlayerTag it has become even more difficult due to nonsensical and unexplained snippets such as:
Why <player.name> and not <PlayerTag.name>?`

Part of why the meta docs were changed was because of exactly that situation - if you don't know what `p@player` or anything else is, you'll try to do `<p@player.name>` ... if you have ever had the player named `player` on your server, this will actually form a valid tag, and work without error. That's an unlikely case, but other cases are more likely (like `n@npc`) and some are *guaranteed* (like `el@element`). `PlayerTag` notation by contrast will always produce an error message... even opening the option to make a specific customized error message for when you use a tag like that (hasn't been done yet but is on my ideas list).

**Comment by Morphan1** [comment]
Thanks for your contribution to this issue, your point of view is helpful to figure out what went wrong. The decision seems to be akin to duct tape to patch user error... where the user error is likely caused by insufficient or otherwise confusing documentation.

As for a possible alternative solution, to my knowledge mechanisms do not have this particular issue despite having similar requirements. Perhaps tag meta could be modified closer to mechanism meta, at least in terms of serving it.

Either way, proper structured documentation in these docs should help alleviate some of those beginner problems. Perhaps this deserves a new issue post, but it may need to have service-specific usage information, such as how to put together UnizenBot meta.

**Comment by mcmonkey4eva** [comment]
Mechanisms do actually have a very similar issue - new users don't know how to turn mechanism meta into something usable, and have to be directed to "how to use mechanisms" tutorials.

**Comment by Morphan1** [comment]
That seems to be an entirely separate non-issue. The issue you brought up with tags is that users would copy and attempt to use the full literal tag. There's no clear way at all to do anything of the sort with mechanisms,  which leads to searching for information rather than blind trial and (if not unlucky) error. The most you can do is clearly provide that information.

**Comment by mcmonkey4eva** [comment]
Fair point. I believe the specialized error message suffices as a counter to that, at least insofar as alerting users to go looking for information (and even giving the reference of what to look for and where to look for it), with only a small bit longer of a delay between user reading the meta and user knowing to go look for info (compared to mechs).
It *could* be changed to displaying the object and the tag name separately, but I think that risks ending up *less* clear ... and generally, more messy formatting. Especially given the tags that aren't in an object->tag format (like `util` tags) which would end up needing their own special different handling, which is both added complexity to the meta handling and added confusion for users looking it.
